# Notes to Improve Workflow

In [127]:
# Setup matplotlib to plot inline (within the notebook)
%matplotlib inline
# Import the pyplot module of Matplotlib as plt
import matplotlib.pyplot as plt
# Import pandas under the abbreviation 'pd'
import pandas as pd
# Import NumPy under the abbreviation 'np'
import numpy as np

#Notes for efficienccy:
#shift + tab when cursor is in arguments to bring up its documentation
#ex: pd.DataFrame(Shift+tab here)
#recall: Json == Dictionary
#to know functions of pandas:
dir(pd)

['BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_hashtable',
 '_is_numpy_dev',
 '_lib',
 '_libs',
 '_np_version_under1p18',
 '_testing',
 '_tslib',
 '_typing',
 

# Reading The Data

In [68]:
# using pandas and read_csv()
acquis = pd.read_csv("datasets/startups/acquisitions.csv")
additions = pd.read_csv("datasets/startups/additions.csv")
companies = pd.read_csv("datasets/startups/companies.csv")
invests = pd.read_csv("datasets/startups/investments.csv")
rounds = pd.read_csv("datasets/startups/rounds.csv")
dfAll = [acquis, additions, companies, invests, rounds]

In [69]:
#shapes (rows x cols) of each dataset
for ele in dfAll:
    print(ele.shape)

(18969, 18)
(2213, 4)
(66368, 14)
(168647, 18)
(114949, 12)


# Samples of the datasets

In [70]:
acquis.sample(5).transpose()

12523  \
company_permalink                                  /organization/paydiant   
company_name                                                     Paydiant   
company_category_list                     Mobile Commerce|Mobile Payments   
company_country_code                                                  USA   
company_state_code                                                     MA   
company_region                                                     Boston   
company_city                                                    Wellesley   
acquirer_permalink                                   /organization/paypal   
acquirer_name                                                      PayPal   
acquirer_category_list  E-Commerce|Finance|FinTech|Mobile Payments|P2P...   
acquirer_country_code                                                 USA   
acquirer_state_code                                                    CA   
acquirer_region                                               SF Bay Area   
acquirer_city                                                    San Jose   
acquired_at                                                      3/2/2015   
acquired_month                                                    2015-03   
price_amount                                                  280000000.0   
price_currency_code                                                   USD   

                                                                9479   \
company_permalink                     /organization/lexicon-marketing   
company_name                                        Lexicon Marketing   
company_category_list   Development Platforms|Language Learning|Video   
company_country_code                                              NaN   
company_state_code                                                NaN   
company_region                                                    NaN   
company_city                                                      NaN   
acquirer_permalink                  /organization/golden-gate-capital   
acquirer_name                                     Golden Gate Capital   
acquirer_category_list                                            NaN   
acquirer_country_code                                             USA   
acquirer_state_code                                                CA   
acquirer_region                                           SF Bay Area   
acquirer_city                                           San Francisco   
acquired_at                                                 1/30/2003   
acquired_month                                                2003-01   
price_amount                                                      NaN   
price_currency_code                                               USD   

                                                   4054   \
company_permalink               /organization/cope-today   
company_name                                  Cope Today   
company_category_list                                NaN   
company_country_code                                 USA   
company_state_code                                    NC   
company_region                                   Raleigh   
company_city                                     Raleigh   
acquirer_permalink      /organization/mindcare-solutions   
acquirer_name                         MindCare Solutions   
acquirer_category_list          Health Services Industry   
acquirer_country_code                                USA   
acquirer_state_code                                   TN   
acquirer_region                                Nashville   
acquirer_city                                  Nashville   
acquired_at                                    11/1/2013   
acquired_month                                   2013-11   
price_amount                                         NaN   
price_currency_code                                  USD   

                                            11051  \
company_permalink     

In [71]:
additions.sample(5).transpose()

900           65            104  \
content    Profiles: Organizations  Acquisitions  Acquisitions   
month_str                  2012-05       1996-08       1999-11   
year_str                      2012          1996          1999   
value                         2237             4            19   

                               911       781  
content    Profiles: Organizations  Profiles  
month_str                  2013-04   2011-02  
year_str                      2013      2011  
value                         8020      4959

In [72]:
companies.sample(5).transpose()

40127                          39531  \
permalink          /organization/nutzvieh24  /organization/node-management   
name                             Nutzvieh24                Node Management   
homepage_url           http://www.vieh24.de                            NaN   
category_list                       Finance                       Software   
funding_total_usd                         -                         960000   
status                               closed                      operating   
country_code                            DEU                            NaN   
state_code                                2                            NaN   
region                          DEU - Other                            NaN   
city                               Miesbach                            NaN   
funding_rounds                            1                              1   
founded_at                              NaN                            NaN   
first_funding_at                 10/13/2010                      3/19/2005   
last_funding_at                  10/13/2010                      3/19/2005   

                                                       38191  \
permalink                      /organization/napera-networks   
name                                         Napera Networks   
homepage_url                           http://www.napera.com   
category_list      Networking|Network Security|SaaS|Security   
funding_total_usd                                    6600000   
status                                              acquired   
country_code                                             USA   
state_code                                                WA   
region                                               Seattle   
city                                                Bellevue   
funding_rounds                                             2   
founded_at                                         12/1/2006   
first_funding_at                                  11/29/2008   
last_funding_at                                     9/2/2009   

                                          21909                      32096  
permalink             /organization/funky-moves  /organization/letsventure  
name                                Funky Moves                LetsVenture  
homepage_url       http://www.funky-moves.co.uk     http://letsventure.com  
category_list                             Games                   Internet  
funding_total_usd                        520094                     650000  
status                                operating                  operating  
country_code                                GBR                        SGP  
state_code                                   W9                        NaN  
region                               Livingston                  Singapore  
city                                 Livingston                  Singapore  
funding_rounds                                1                          2  
founded_at                                  NaN                  9/13/2013  
first_funding_at                      5/29/2012                  5/13/2014  
last_funding_at                       5/29/2012                 10/19/2015

In [73]:
invests.sample(5).transpose()

159548  \
company_permalink                                      /organization/vserv   
company_name                                                         Vserv   
company_category_list    Emerging Markets|Mobile Advertising|Mobile Com...   
company_country_code                                                   IND   
company_state_code                                                      16   
company_region                                                      Mumbai   
company_city                                                        Mumbai   
investor_permalink                         /organization/epiphany-ventures   
investor_name                                            Epiphany Ventures   
investor_country_code                                                  IND   
investor_state_code                                                     16   
investor_region                                                     Mumbai   
investor_city                                                       Mumbai   
funding_round_permalink    /funding-round/2ca06bccd055014b699b896d5585b82f   
funding_round_type                                                 venture   
funding_round_code                                                       B   
funded_at                                                         9/1/2012   
raised_amount_usd                                                4000000.0   

                                                                    161148  \
company_permalink                             /organization/weddington-way   
company_name                                                Weddington Way   
company_category_list    Consumer Internet|E-Commerce|Events|Fashion|So...   
company_country_code                                                   USA   
company_state_code                                                      CA   
company_region                                                 SF Bay Area   
company_city                                                 San Francisco   
investor_permalink                          /organization/felicis-ventures   
investor_name                                             Felicis Ventures   
investor_country_code                                                  USA   
investor_state_code                                                     CA   
investor_region                                                SF Bay Area   
investor_city                                                    Palo Alto   
funding_round_permalink    /funding-round/958357e1ac39fa023078c49795e93687   
funding_round_type                                                    seed   
funding_round_code                                                     NaN   
funded_at                                                         8/1/2011   
raised_amount_usd                                                2500000.0   

                                                                  23504   \
company_permalink                       /organization/bubbles-and-beyond   
company_name                                          Bubbles and Beyond   
company_category_list                                      Biotechnology   
company_country_code                                                 DEU   
company_state_code                                                    13   
company_region                                                   Leipzig   
company_city                                                     Leipzig   
investor_permalink                 /organization/high-tech-gruenderfonds   
investor_name                                    High-Tech Gruenderfonds   
investor_country_code                                                DEU   
investor_state_code                                                    7   
investor_region                                                     Bonn   
investor_city                                                       Bonn   
funding_round_permalink  /funding-roun

In [74]:
rounds.sample(5).transpose()

31892  \
company_permalink                                  /organization/enertiv   
company_name                                                     Enertiv   
company_category_list                        Analytics|Energy Efficiency   
company_country_code                                                 NaN   
company_state_code                                                   NaN   
company_region                                                       NaN   
company_city                                                         NaN   
funding_round_permalink  /funding-round/79627fdd6199f0567dc37393975d95b7   
funding_round_type                                                  seed   
funding_round_code                                                   NaN   
funded_at                                                       1/1/2011   
raised_amount_usd                                               120000.0   

                                                                   4865   \
company_permalink                                   /organization/alterg   
company_name                                                      AlterG   
company_category_list                        Health Care|Medical Devices   
company_country_code                                                 USA   
company_state_code                                                    CA   
company_region                                               SF Bay Area   
company_city                                                     Fremont   
funding_round_permalink  /funding-round/527a8af29fabe1c6d0a49422dd74aeac   
funding_round_type                                               venture   
funding_round_code                                                   NaN   
funded_at                                                       8/8/2012   
raised_amount_usd                                             10310133.0   

                                                                   1490   \
company_permalink                                   /organization/accera   
company_name                                                      Accera   
company_category_list                                      Biotechnology   
company_country_code                                                 USA   
company_state_code                                                    CO   
company_region                                                    Denver   
company_city                                                  Broomfield   
funding_round_permalink  /funding-round/e6658050ba8d6d2e23060ca6e7ae8d92   
funding_round_type                                               venture   
funding_round_code                                                     C   
funded_at                                                     11/12/2008   
raised_amount_usd                                             35000000.0   

                                                                   65747  \
company_permalink                                   /organization/nanomr   
company_name                                                      nanoMR   
company_category_list                                      Biotechnology   
company_country_code                                                 USA   
company_state_code                                                    NM   
company_region                                               Albuquerque   
company_city                                                 Albuquerque   
funding_round_permalink  /funding-round/8e794b80ff062c4eb923adf09ff713df   
funding_round_type                                        debt_financing   
funding_round_code                                                   NaN   
funded_at                                                       7/9/2013   
raised_amount_usd                                              2623512.0   

                                                                   76552  
company_permalink                   

In [75]:
#getting all (unique) columns of the datasets
uniqueCols = set() #or: instead of "{}", "set()"
for ele in dfAll:
    cols = ele.columns.tolist() #".tolist()" to make it a list (side note: ".values" to make it a numpy.ndarray)
    
    uniqueCols.update(cols) #use ".add()" for elements, ".update()" for lists 

In [76]:
#sorting the columns alphabetically
tempList = list(uniqueCols)
tempList.sort()
uniqueCols = tempList
uniqueCols

['acquired_at',
 'acquired_month',
 'acquirer_category_list',
 'acquirer_city',
 'acquirer_country_code',
 'acquirer_name',
 'acquirer_permalink',
 'acquirer_region',
 'acquirer_state_code',
 'category_list',
 'city',
 'company_category_list',
 'company_city',
 'company_country_code',
 'company_name',
 'company_permalink',
 'company_region',
 'company_state_code',
 'content',
 'country_code',
 'first_funding_at',
 'founded_at',
 'funded_at',
 'funding_round_code',
 'funding_round_permalink',
 'funding_round_type',
 'funding_rounds',
 'funding_total_usd',
 'homepage_url',
 'investor_city',
 'investor_country_code',
 'investor_name',
 'investor_permalink',
 'investor_region',
 'investor_state_code',
 'last_funding_at',
 'month_str',
 'name',
 'permalink',
 'price_amount',
 'price_currency_code',
 'raised_amount_usd',
 'region',
 'state_code',
 'status',
 'value',
 'year_str']

# Seeing the columns in the datasets

In [77]:
def datasetsCols(dfList, datasetNames, pad=""):
    """
    Takes dataframes and returns a dataframe (df) with each column having a df's
    column names. The "pad" is to make sure all lists of columns have 
    the same length padded out with "pad"
    Example
    pad = "XX", datasetNames = ["cars", "planes"]
    dfList = [carsDF, planesDF]
    returned data frame:
            cars            planes
    0       numOfWheels     numOfWings
    1       manufacturer    manufacturer
    2       make            XX
    """
    cols = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        dictCsvs.update({datasetNames[i] : cols[i]})

    return pd.DataFrame(dictCsvs)

In [78]:
lstCsvNames = ["acquisitions", "additions", "companies", "investments", "rounds"]

dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions          companies  \
0        company_permalink    content          permalink   
1             company_name  month_str               name   
2    company_category_list   year_str       homepage_url   
3     company_country_code      value      category_list   
4       company_state_code             funding_total_usd   
5           company_region                        status   
6             company_city                  country_code   
7       acquirer_permalink                    state_code   
8            acquirer_name                        region   
9   acquirer_category_list                          city   
10   acquirer_country_code                funding_rounds   
11     acquirer_state_code                    founded_at   
12         acquirer_region              first_funding_at   
13           acquirer_city               last_funding_at   
14             acquired_at                                 
15          acquired_month                                 
16            price_amount                                 
17     price_currency_code                                 

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

# Data Cleaning

First, lets change the columns in the `companies` dataset <br>
to be the similar to all other datasets:

In [79]:
companies.rename(columns = {'permalink' : 'company_permalink', 
                            'name' : 'company_name',
                            'category_list' : 'company_category_list',
                            'country_code' : 'company_country_code',
                            'state_code' : 'company_state_code',
                            'region' : 'company_region',
                            'city' : 'company_city'}, inplace = True)

In [80]:
dfAll[2] = companies
dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

Let's start by analyzing `additions` dataset, as it has unusual format of columns

## Removing "additions" Dataset

In [81]:
additions.sample(5)

content month_str  year_str  value
1390  Rounds: convertible_note   2014-01      2014     42
898    Profiles: Organizations   2012-03      2012   2069
275               Acquisitions   2014-02      2014    192
1764    Rounds: private_equity   2009-02      2009      2
1821    Rounds: private_equity   2013-12      2013     37

In [82]:
additions["content"].unique()

array(['Acquisitions', 'IPOs', 'Profiles', 'Profiles: Organizations',
       'Profiles: People', 'Profiles: Products', 'Rounds',
       'Rounds: angel', 'Rounds: convertible_note',
       'Rounds: debt_financing', 'Rounds: equity_crowdfunding',
       'Rounds: grant', 'Rounds: non_equity_assistance',
       'Rounds: post_ipo_debt', 'Rounds: post_ipo_equity',
       'Rounds: private_equity', 'Rounds: product_crowdfunding',
       'Rounds: secondary_market', 'Rounds: seed', 'Rounds: undisclosed',
       'Rounds: venture'], dtype=object)

In [83]:
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

In [84]:
additions.shape

(2213, 4)

In [85]:
companies.shape, rounds.shape

((66368, 14), (114949, 12))

So apparently there are no columns that could be used to join with the other datasets. <br>
Furthermore, the "value" column is too vague to be useful. <br>
Therefore, the "additions" dataset will be discarded.

In [86]:
dfAll.pop(1)
lstCsvNames.pop(1)
len(dfAll), lstCsvNames

(4, ['acquisitions', 'companies', 'investments', 'rounds'])

# Data Cleaning (Cont.)

## Checking All Missing Values

In [87]:
def datasetsNulls(dfList, datasetNames, nullCol="nulls_", pad=""):
    """
    Use this when you want to display a column of column names,
    then a column of the null values, and repeat that for each dataset.
    Returns a dataframe
    """
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.isnull().sum().tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[nullCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [88]:
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        1           company_name      1.0   
2    company_category_list     4984           homepage_url   5058.0   
3     company_country_code     4669  company_category_list   3148.0   
4       company_state_code     4949      funding_total_usd      0.0   
5           company_region     4911                 status      0.0   
6             company_city     4911   company_country_code   6958.0   
7       acquirer_permalink        1     company_state_code   8547.0   
8            acquirer_name        1         company_region   8030.0   
9   acquirer_category_list     2389           company_city   8028.0   
10   acquirer_country_code     1171         funding_rounds      0.0   
11     acquirer_state_code     1436             founded_at  15221.0   
12         acquirer_region     1360       first_funding_at     24.0   
13           acquirer_city     1360        last_funding_at      0.0   
14             acquired_at       13                             NaN   
15          acquired_month       15                             NaN   
16            price_amount    13957                             NaN   
17     price_currency_code        7                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        1             company_name      1.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink       12  funding_round_permalink      0.0  
8             investor_name       12       funding_round_type      0.0  
9     investor_country_code    30939       funding_round_code  83809.0  
10      investor_state_code    44739                funded_at      0.0  
11          investor_region    43217        raised_amount_usd  19990.0  
12            investor_city    32164                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87585                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

Alternatively, we can combine all columns to be the dataframe's index:

In [89]:
#Alternatively:
def datasetsNullsSameIndex(datasetsNames, *args):
    """
    Use this when multiple datasets have a lot of columns
    with the same names.
    Returns a dataframe with the index value being the 
    columns of all datasets
    """
    dic = {}
    for i, df in enumerate(args):
        dic[datasetsNames[i]] = df.isnull().sum()
    return pd.DataFrame(dic)

In [90]:
datasetsNullsSameIndex(lstCsvNames, *dfAll)

acquisitions  companies  investments   rounds
acquired_at                      13.0        NaN          NaN      NaN
acquired_month                   15.0        NaN          NaN      NaN
acquirer_category_list         2389.0        NaN          NaN      NaN
acquirer_city                  1360.0        NaN          NaN      NaN
acquirer_country_code          1171.0        NaN          NaN      NaN
acquirer_name                     1.0        NaN          NaN      NaN
acquirer_permalink                1.0        NaN          NaN      NaN
acquirer_region                1360.0        NaN          NaN      NaN
acquirer_state_code            1436.0        NaN          NaN      NaN
company_category_list          4984.0     3148.0       3632.0   3410.0
company_city                   4911.0     8028.0      12483.0  10164.0
company_country_code           4669.0     6958.0      10521.0   8678.0
company_name                      1.0        1.0          1.0      1.0
company_permalink                 0.0        0.0          0.0      0.0
company_region                 4911.0     8030.0      12485.0  10167.0
company_state_code             4949.0     8547.0      13431.0  10946.0
first_funding_at                  NaN       24.0          NaN      NaN
founded_at                        NaN    15221.0          NaN      NaN
funded_at                         NaN        NaN          0.0      0.0
funding_round_code                NaN        NaN      87585.0  83809.0
funding_round_permalink           NaN        NaN          0.0      0.0
funding_round_type                NaN        NaN          0.0      0.0
funding_rounds                    NaN        0.0          NaN      NaN
funding_total_usd                 NaN        0.0          NaN      NaN
homepage_url                      NaN     5058.0          NaN      NaN
investor_city                     NaN        NaN      32164.0      NaN
investor_country_code             NaN        NaN      30939.0      NaN
investor_name                     NaN        NaN         12.0      NaN
investor_permalink                NaN        NaN         12.0      NaN
investor_region                   NaN        NaN      43217.0      NaN
investor_state_code               NaN        NaN      44739.0      NaN
last_funding_at                   NaN        0.0          NaN      NaN
price_amount                  13957.0        NaN          NaN      NaN
price_currency_code               7.0        NaN          NaN      NaN
raised_amount_usd                 NaN        NaN      21619.0  19990.0
status                            NaN        0.0          NaN      NaN

### `dropna()` for columns with few missing values

Notice that there is one row that doesn't contain a company name, <br>
so let's validate that it is the same company across all datasets, <br>
so we can remove it:

In [91]:
datasetsCols(dfAll, lstCsvNames)

acquisitions              companies              investments  \
0        company_permalink      company_permalink        company_permalink   
1             company_name           company_name             company_name   
2    company_category_list           homepage_url    company_category_list   
3     company_country_code  company_category_list     company_country_code   
4       company_state_code      funding_total_usd       company_state_code   
5           company_region                 status           company_region   
6             company_city   company_country_code             company_city   
7       acquirer_permalink     company_state_code       investor_permalink   
8            acquirer_name         company_region            investor_name   
9   acquirer_category_list           company_city    investor_country_code   
10   acquirer_country_code         funding_rounds      investor_state_code   
11     acquirer_state_code             founded_at          investor_region   
12         acquirer_region       first_funding_at            investor_city   
13           acquirer_city        last_funding_at  funding_round_permalink   
14             acquired_at                              funding_round_type   
15          acquired_month                              funding_round_code   
16            price_amount                                       funded_at   
17     price_currency_code                               raised_amount_usd   

                     rounds  
0         company_permalink  
1              company_name  
2     company_category_list  
3      company_country_code  
4        company_state_code  
5            company_region  
6              company_city  
7   funding_round_permalink  
8        funding_round_type  
9        funding_round_code  
10                funded_at  
11        raised_amount_usd  
12                           
13                           
14                           
15                           
16                           
17

In [92]:
#company_permalink of each row that doesn't have a company name:
for df in dfAll:
    print(df[df['company_name'].isnull()]['company_permalink'].values)

['mendatio"']
['/organization/tell-it-in']
['/organization/tell-it-in']
['/organization/tell-it-in']


So apparently, the `acquisitions` dataset has a different permalink that has no company name, <br>
So we could remove these two companies, as the total number of companies are big enough to get the insight that we want: 

In [93]:
for df in dfAll:
    df.dropna(subset=['company_name'], inplace=True)
#Check:
for df in dfAll:
    print(df[df['company_name'].isnull()]['company_permalink'].values)

[]
[]
[]
[]


`dfAllNulls` dataframe also shows that there are missing values for `acquirer_name` and `investor_name` <br>
which can't be imputed, so we'll also remove them:

In [94]:
# Note: same logic applies for the permalinks
acquis.dropna(subset=['acquirer_permalink', 'acquirer_name'], inplace=True)
invests.dropna(subset=['investor_permalink', 'investor_name'], inplace=True)

Also, we'll remove the missing values for columns that have less than 100 `NaN`, <br>
as they're small in comparison to their respective datasets:

In [95]:
for df in dfAll:
    nulls = df.isnull().sum()
    cols = nulls[(nulls <= 100) & (nulls != 0)].index.tolist()
    df.dropna(subset=cols, inplace=True)
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        0           company_name      0.0   
2    company_category_list     4977           homepage_url   5058.0   
3     company_country_code     4661  company_category_list   3147.0   
4       company_state_code     4940      funding_total_usd      0.0   
5           company_region     4902                 status      0.0   
6             company_city     4902   company_country_code   6956.0   
7       acquirer_permalink        0     company_state_code   8545.0   
8            acquirer_name        0         company_region   8028.0   
9   acquirer_category_list     2387           company_city   8026.0   
10   acquirer_country_code     1168         funding_rounds      0.0   
11     acquirer_state_code     1432             founded_at  15218.0   
12         acquirer_region     1356       first_funding_at      0.0   
13           acquirer_city     1356        last_funding_at      0.0   
14             acquired_at        0                             NaN   
15          acquired_month        0                             NaN   
16            price_amount    13941                             NaN   
17     price_currency_code        0                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        0             company_name      0.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink        0  funding_round_permalink      0.0  
8             investor_name        0       funding_round_type      0.0  
9     investor_country_code    30927       funding_round_code  83808.0  
10      investor_state_code    44727                funded_at      0.0  
11          investor_region    43205        raised_amount_usd  19990.0  
12            investor_city    32152                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87576                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

## Converting & Imputing Dates

[SAY WHAT YOU WILL DO HERE FARAH:]

In [96]:
#Farah: plz impute (not drop) the columns' missing values related to dates in each dataset which are:
#founded_at.. and that's it :] 
# Suggestion: convert the dates of all the columns that have date values from "object" to datetime
#       Helpful link: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#converting-to-timestamps
# Then, make the founded_at a couple of days before first_funding_at

#Sincerely No. -Farah :)

In [97]:
#farah stop point here

## Cleaning & Imputing Geographical Locations

### Removing `company_state_code` and `company_region`

We don't need `company_state_code` in our analysis, <br>
As country code and city are sufficient to know <br>
an approximation of the geographical location of the company. <br>
Therefore, let's remove it:

In [98]:
companies[['company_country_code', 'company_state_code', 'company_city']].head(10)

company_country_code company_state_code   company_city
0                  IND                 16         Mumbai
1                  USA                 DE  Delaware City
2                  NaN                NaN            NaN
3                  CHN                 22        Beijing
4                  USA                 IL      Champaign
5                  HKG                NaN      Hong Kong
6                  CAN                 BC      Vancouver
7                  USA                 CA  Mountain View
8                  USA                 CA  San Francisco
9                  CHL                 12     Las Condes

In [99]:
companies.drop('company_state_code', axis=1, inplace=True)

`company_region` is also redundent, as it is usually the same as `company_city`.<br>
Let's validate this:

In [118]:
companies[companies.company_region == companies.company_city][['company_region', 'company_city']]

company_region company_city
0             Mumbai       Mumbai
3            Beijing      Beijing
5          Hong Kong    Hong Kong
6          Vancouver    Vancouver
11           Orlando      Orlando
...              ...          ...
66354           Bath         Bath
66360      Milwaukee    Milwaukee
66362         London       London
66363        Beijing      Beijing
66364          Split        Split

[25360 rows x 2 columns]

Let's see the mean of similarity between the strings using builtin library `difflib.SequenceMatcher`:

In [123]:
from difflib import SequenceMatcher
similarities = []
rgs = companies.company_region.values.tolist()
cts = companies.company_city.values.tolist()
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]): # If either of the elements is NaN, then don't find the similarity
        similarities.append(SequenceMatcher(None, rgs[i], cts[i]).ratio()) # Gets the similarity between each two strings
avgSim = sum(similarities) / len(similarities)
avgSim

0.6175093624179544

However, `SequenceMatcher` only finds the longest contiguous matching subsequence, <br>
and all other characters will drastically decrease the similarity, which isn't very accurate, <br>
Therefore, let's try to see how many times a string is a substring of another:

In [124]:
subsCount = 0
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]):
        subsCount += (rgs[i] in cts[i] or cts[i] in rgs[i])
subsCount

29884

So to summarize: almost half of the companies have the same data for `region` and `city` (25360) <br>
There are around 5000 companies that have the same semantic meaning between the data <br>
(e.g: 'New York City' is the same as 'NYC, New York City') <br>
Therefore, `company_region` can be removed without losing any possible future insights:

In [125]:
companies.drop('company_region', axis=1, inplace=True)

### Imputing `company_country_code` and `company_city`

A way to impute country codes is to do the following:
1. check the company's permalink
2. find HTML that contains 

In [100]:
companies[companies.company_country_code.isnull() == False].sample(10)

company_permalink          company_name  \
59235           /organization/trending-taste        Trending Taste   
38588                  /organization/neozone               Neozone   
60491     /organization/united-digital-group  United Digital Group   
34370  /organization/marquee-productions-inc                   MPI   
60156                 /organization/uber-com              Uber.com   
2138                    /organization/aireon                Aireon   
53134           /organization/solar-universe        Solar Universe   
39344                    /organization/nifti                 Nifti   
28210      /organization/integrys-assetpoint   Integrys AssetPoint   
23730            /organization/gozaround-inc        GozAround Inc.   

                                            homepage_url  \
59235                          http://trendingtaste.com/   
38588                             http://www.neozone.com   
60491                                 http://www.udg.de/   
34370                        http://www.marqueeworld.com   
60156  http://www.dmwmedia.com/news/2008/09/29/social...   
2138                               http://www.aireon.com   
53134                           http://solaruniverse.com   
39344                                   http://nifti.com   
28210                          http://www.assetpoint.com   
23730                           http://www.gozaround.com   

                                   company_category_list funding_total_usd  \
59235                                Entertainment|Games                 -   
38588                                           Software            809000   
60491                                        Advertising                 -   
34370                                              Music                 -   
60156  Communications Infrastructure|Media|Social Net...           7600000   
2138                               Public Transportation          50000000   
53134  Clean Energy|Clean Technology|Franchises|Renew...          11000000   
39344                                        Curated Web            800000   
28210   Business Services|Productivity Software|Software           5500000   
23730            Non Profit|Social Commerce|Social Media             46000   

          status company_country_code            company_region company_city  \
59235  operating                  USA               SF Bay Area     San Jose   
38588  operating                  DNK                Copenhagen   Copenhagen   
60491  operating                  DEU                   Hamburg      Hamburg   
34370  operating                  USA               Los Angeles       Malibu   
60156     closed                  USA               Los Angeles  Los Angeles   
2138   operating                  USA          Washington, D.C.       Mclean   
53134  operating                  USA               SF Bay Area    Livermore   
39344  operating                  USA                    Boston       Boston   
28210  operating                  USA  Greenville - Spartanburg   Greenville   
23730  operating                  CAN                  Edmonton     Edmonton   

       funding_rounds founded_at first_funding_at last_funding_at  
59235               1  6/22/2013         7/2/2014        7/2/2014  
38588               1        NaN        1/17/2006       1/17/2006  
60491               1        NaN         6/1/2011        6/1/2011  
34370               1  10/4/2010       10/10/2010      10/10/2010  
60156               1        NaN        5/26/2008       5/26/2008  
2138                1        NaN        2/18/2014       2/18/2014  
53134               2   1/1/2007        1/25/2011       4/20/2015  
39344               1   1/1/2012        7/18/2013       7/18/2013  
28210               1   1/1/2002         3/6/2008        3/6/2008  
23730               1  1/15/2014        1/15/2014       1/15/2014

In [101]:
companies[companies.company_country_code.isnull() == False].sample(5)

company_permalink            company_name  \
50151  /organization/seattle-coffee-company  Seattle Coffee Company   
12681              /organization/compassinc                 Compass   
36208        /organization/mission-research        Mission Research   
10551       /organization/cetek-corporation       Cetek Corporation   
12807                /organization/concloud                ConCloud   

                   homepage_url  \
50151                       NaN   
12681    http://www.compass.com   
36208  http://www.circledog.com   
10551     http://www.cetek.com/   
12807   http://www.concloud.com   

                                   company_category_list funding_total_usd  \
50151                  Coffee|E-Commerce|Food Processing                 -   
12681               Curated Web|Local|Mobile|Real Estate         123000000   
36208                                     Nanotechnology           1500000   
10551                                          Bio-Pharm          20000000   
12807  Android|Apps|Cloud Computing|Contact Managemen...            100000   

          status company_country_code company_region company_city  \
50151  operating                  USA  New York City      Astoria   
12681  operating                  USA  New York City     New York   
36208   acquired                  USA     Harrisburg    Lancaster   
10551  operating                  USA         Boston  Marlborough   
12807  operating                  DEU      Frankfurt    Frankfurt   

       funding_rounds founded_at first_funding_at last_funding_at  
50151               1   2/1/2014        2/10/2013       2/10/2013  
12681               4  10/1/2012       12/16/2012       9/15/2015  
36208               2  2/15/2002       11/30/2006       2/25/2008  
10551               2   1/1/1996       10/12/2001        4/2/2003  
12807               1  2/28/2014        2/28/2014       2/28/2014

In [102]:
companies.company_country_code.unique()

array(['IND', 'USA', nan, 'CHN', 'HKG', 'CAN', 'CHL', 'GBR', 'FRA', 'AUS',
       'DNK', 'ROM', 'AUT', 'KOR', 'NLD', 'SWE', 'JPN', 'RUS', 'SGP',
       'NOR', 'COL', 'ESP', 'BEL', 'IRL', 'ITA', 'ISR', 'NZL', 'CZE',
       'DEU', 'CHE', 'BRA', 'BGR', 'SVN', 'JOR', 'HUN', 'BWA', 'NGA',
       'FIN', 'TUR', 'IDN', 'ARE', 'CYP', 'POL', 'CRI', 'PRT', 'ARG',
       'TWN', 'KHM', 'THA', 'SVK', 'UKR', 'LTU', 'ISL', 'ZAF', 'MEX',
       'VEN', 'URY', 'MUS', 'KEN', 'PHL', 'MNE', 'VNM', 'GHA', 'PSE',
       'MYS', 'PER', 'EGY', 'PAN', 'LVA', 'GGY', 'ALB', 'UGA', 'HRV',
       'EST', 'LBN', 'GRC', 'PAK', 'NPL', 'LUX', 'SAU', 'IRN', 'GTM',
       'BAH', 'ARM', 'BGD', 'MDA', 'SRB', 'TUN', 'ZWE', 'BRB', 'NIC',
       'TAN', 'PRI', 'TTO', 'BHR', 'CMR', 'BLR', 'CYM', 'BRN', 'ECU',
       'SLV', 'MLT', 'ZMB', 'MKD', 'GIB', 'TGO', 'LAO', 'BMU', 'HND',
       'MCO', 'MMR', 'KAZ', 'QAT', 'DOM', 'MAR', 'LIE', 'GEO', 'MOZ',
       'DZA', 'JAM', 'KNA', 'AZE', 'SEN', 'RWA', 'DMA', 'UZB', 'BLZ',
       'OMN', '

## Checking All `dtypes`

In [103]:
def datasetsDtypes(dfList, datasetNames, dtypesCol="dtypes_", pad=""):
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.dtypes.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[dtypesCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [104]:
datasetsDtypes(dfAll, lstCsvNames)

acquisitions dtypes_1              companies dtypes_2  \
0        company_permalink   object      company_permalink   object   
1             company_name   object           company_name   object   
2    company_category_list   object           homepage_url   object   
3     company_country_code   object  company_category_list   object   
4       company_state_code   object      funding_total_usd   object   
5           company_region   object                 status   object   
6             company_city   object   company_country_code   object   
7       acquirer_permalink   object         company_region   object   
8            acquirer_name   object           company_city   object   
9   acquirer_category_list   object         funding_rounds    int64   
10   acquirer_country_code   object             founded_at   object   
11     acquirer_state_code   object       first_funding_at   object   
12         acquirer_region   object        last_funding_at   object   
13           acquirer_city   object                            None   
14             acquired_at   object                            None   
15          acquired_month   object                            None   
16            price_amount  float64                            None   
17     price_currency_code   object                            None   

                investments dtypes_3                   rounds dtypes_4  
0         company_permalink   object        company_permalink   object  
1              company_name   object             company_name   object  
2     company_category_list   object    company_category_list   object  
3      company_country_code   object     company_country_code   object  
4        company_state_code   object       company_state_code   object  
5            company_region   object           company_region   object  
6              company_city   object             company_city   object  
7        investor_permalink   object  funding_round_permalink   object  
8             investor_name   object       funding_round_type   object  
9     investor_country_code   object       funding_round_code   object  
10      investor_state_code   object                funded_at   object  
11          investor_region   object        raised_amount_usd  float64  
12            investor_city   object                              None  
13  funding_round_permalink   object                              None  
14       funding_round_type   object                              None  
15       funding_round_code   object                              None  
16                funded_at   object                              None  
17        raised_amount_usd  float64                              None

In [3]:
#AISHA: see what you can do with the dtypes above (except dates, that's on Farah) ^^

testing something here


In [1]:
print("Aisha is happy ^^")

Aisha is happy ^^


In [106]:
#Aisha stop

## `rounds` Dataset

In [107]:
dfAllCols.iloc[:, -2:]

investments                   rounds
0         company_permalink        company_permalink
1              company_name             company_name
2     company_category_list    company_category_list
3      company_country_code     company_country_code
4        company_state_code       company_state_code
5            company_region           company_region
6              company_city             company_city
7        investor_permalink  funding_round_permalink
8             investor_name       funding_round_type
9     investor_country_code       funding_round_code
10      investor_state_code                funded_at
11          investor_region        raised_amount_usd
12            investor_city                         
13  funding_round_permalink                         
14       funding_round_type                         
15       funding_round_code                         
16                funded_at                         
17        raised_amount_usd

In [108]:
invests.shape, rounds.shape

((168634, 18), (114948, 12))

In [109]:
invests.head()

company_permalink        company_name company_category_list  \
0             /organization/0-6-com             0-6.com           Curated Web   
1    /organization/004-technologies    004 Technologies              Software   
2  /organization/01games-technology  01Games Technology                 Games   
3              /organization/0xdata              H2O.ai             Analytics   
4              /organization/0xdata              H2O.ai             Analytics   

  company_country_code company_state_code         company_region  \
0                  CHN                 22                Beijing   
1                  USA                 IL  Springfield, Illinois   
2                  HKG                NaN              Hong Kong   
3                  USA                 CA            SF Bay Area   
4                  USA                 CA            SF Bay Area   

    company_city                    investor_permalink  \
0        Beijing     /organization/dt-capital-partners   
1      Champaign       /organization/venturecapital-de   
2      Hong Kong     /organization/cyberport-hong-kong   
3  Mountain View             /organization/capital-one   
4  Mountain View  /organization/nexus-venture-partners   

            investor_name investor_country_code investor_state_code  \
0     DT Capital Partners                   CHN                  23   
1   VCDE Venture Partners                   DEU                   5   
2     Cyberport Hong Kong                   HKG                 NaN   
3             Capital One                   USA                  VA   
4  Nexus Venture Partners                   USA                  CA   

    investor_region investor_city  \
0          Shanghai      Shanghai   
1         Frankfurt     Frankfurt   
2         Hong Kong     Hong Kong   
3  Washington, D.C.        Mclean   
4       SF Bay Area    Menlo Park   

                           funding_round_permalink funding_round_type  \
0  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
1  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
2  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
3  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
4  /funding-round/383a9bd2c04f7038bb543ccef5ba3eae               seed   

  funding_round_code  funded_at  raised_amount_usd  
0                  A  3/19/2008          2000000.0  
1                NaN  7/24/2014                NaN  
2                NaN   7/1/2014            41250.0  
3                  B  11/9/2015         20000000.0  
4                NaN  5/22/2013          3000000.0

In [110]:
rounds.head()

company_permalink            company_name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

                               company_category_list company_country_code  \
0                                              Media                  IND   
1  Application Platforms|Real Time|Social Network...                  USA   
2  Application Platforms|Real Time|Social Network...                  USA   
3                                  Apps|Games|Mobile                  NaN   
4                                        Curated Web                  CHN   

  company_state_code company_region   company_city  \
0                 16         Mumbai         Mumbai   
1                 DE     DE - Other  Delaware City   
2                 DE     DE - Other  Delaware City   
3                NaN            NaN            NaN   
4                 22        Beijing        Beijing   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B    1/5/2015         10000000.0  
1                  A  10/14/2014                NaN  
2                NaN    3/1/2014           700000.0  
3                  B   1/30/2014          3406878.0  
4                  A   3/19/2008          2000000.0

In [111]:
len(invests["company_name"].unique()), len(rounds["company_name"].unique())

(44577, 66102)

Since there are companies in `rounds` dataset that are not in `investments` dataset, <br>
therefore we should keep `rounds` dataset for further analysis.

## `companies` Dataset

In [112]:
companies.sample(5).T

64028  \
company_permalink      /organization/windward-naval   
company_name                               Windward   
homepage_url                http://www.windward.eu/   
company_category_list            Big Data Analytics   
funding_total_usd                          15800000   
status                                    operating   
company_country_code                            ISR   
company_region                             Tel Aviv   
company_city                               Tel Aviv   
funding_rounds                                    2   
founded_at                                 1/1/2010   
first_funding_at                          11/7/2013   
last_funding_at                           4/27/2015   

                                              54848  \
company_permalink      /organization/storefront-net   
company_name                         StoreFront.net   
homepage_url              http://www.storefront.net   
company_category_list                    E-Commerce   
funding_total_usd                           6000000   
status                                     acquired   
company_country_code                            USA   
company_region                          Kansas City   
company_city                                 Olathe   
funding_rounds                                    1   
founded_at                                 1/1/1996   
first_funding_at                           2/1/2007   
last_funding_at                            2/1/2007   

                                               36853                    52160  \
company_permalink      /organization/money-dashboard   /organization/skuserve   
company_name                         Money Dashboard                 SkuServe   
homepage_url           http://www.moneydashboard.com  http://www.skuserve.com   
company_category_list   Curated Web|Personal Finance                      NaN   
funding_total_usd                            7791178                        -   
status                                     operating                operating   
company_country_code                             GBR                      USA   
company_region                             Edinburgh                Cleveland   
company_city                               Edinburgh                Cleveland   
funding_rounds                                     2                        1   
founded_at                                  1/1/2008                      NaN   
first_funding_at                          11/28/2013                4/30/2013   
last_funding_at                            1/26/2015                4/30/2013   

                                                   65008  
company_permalink               /organization/yakarouler  
company_name                                  Yakarouler  
homepage_url                  http://www.yakarouler.com/  
company_category_list  Automotive|Discounts|Marketplaces  
funding_total_usd                                3925845  
status                                         operating  
company_country_code                                 FRA  
company_region                               FRA - Other  
company_city                                      Thiais  
funding_rounds                                         1  
founded_at                                      1/1/2006  
first_funding_at                               12/1/2010  
last_funding_at                                12/1/2010

In [113]:
companies.isnull().sum()

company_permalink            0
company_name                 0
homepage_url              5058
company_category_list     3147
funding_total_usd            0
status                       0
company_country_code      6956
company_region            8028
company_city              8026
funding_rounds               0
founded_at               15218
first_funding_at             0
last_funding_at              0
dtype: int64

we will need to remove the none values of the following columns <br>
based on the question that we are trying to gain insight for: <br>
1.	Can the factors that affect a start-up’s growth be determined?
    -	For this: We are initially interested in <br> `funding_total_usd`, `status`, `funding_rounds`, `first_funding_at`, and `last_funding_at`
    -   So we could remove the 24 null values of `first_funding_at` since they are so small <br>
        compared to the rest of the dataset
2.  Do taxes play an important factor in deciding a start-up’s success?

